service.go

```diff
@@ -1047,6 +1056,43 @@ func handleUserDelete(ctx *serviceContext, params []string) error {
 	return nil
 }
 
+func handleUserRun(ctx *serviceContext, params []string) error {
+	if !ctx.user.Admin {
+		return fmt.Errorf("only admins may run command as other users")
+	}
+	if len(params) < 2 {
+		return fmt.Errorf("expected at least two arguments")
+	}
+
+	username := params[0]
+	params = params[1:]
+	if username == ctx.user.Username {
+		handleServiceCommand(ctx, params)
+		return nil
+	}
+
+	u := ctx.user.srv.getUser(username)
+	if u == nil {
+		return fmt.Errorf("unknown username %q", username)
+	}
+
+	printCh := make(chan string, 1)
+	ev := eventUserRun{
+		params: params,
+		print:  printCh,
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case u.events <- ev:
+	}
+	for text := range printCh {
```

Comment:
There is a race here:

- We send `ev` to `u.events`
- The user goroutine stops (because of a crash or user deletion)
- We block on `printCh`

To at least avoid blocking forever the admin goroutine, we can use a `select` when reading from `printCh`, and exit on `ctx.Done()`.

Reply:
(Ideally please add a TODO describing this issue, so that we can fix it properly later, probably by flushing the `u.events` queue and `close(ev.print)` in a `defer`.)